File: problems/leetcode/0000/54.cpp
// 54. Spiral Matrix
// https://leetcode.com/problems/spiral-matrix/description/

#include <vector>

using namespace std;

class Solution {
public:
  vector<int> spiralOrder(vector<vector<int>> &matrix) {
    int m = matrix.size() - 1, n = matrix[0].size();
    vector<int> orderedValues;

    int x = 0, y = -1;
    do {
      if (n <= 0) {
        break;
      }
      for (int i = 1; i <= n; i++) {
        y++;
        orderedValues.push_back(matrix[x][y]);
      }
      n--;

      if (m <= 0) {
        break;
      }
      for (int i = 1; i <= m; i++) {
        x++;
        orderedValues.push_back(matrix[x][y]);
      }
      m--;

      if (n <= 0) {
        break;
      }
      for (int i = 1; i <= n; i++) {
        y--;
        orderedValues.push_back(matrix[x][y]);
      }
      n--;

      if (m <= 0) {
        break;
      }
      for (int i = 1; i <= m; i++) {
        x--;
        orderedValues.push_back(matrix[x][y]);
      }
      m--;
    } while (m > 0 || n > 0);

    return orderedValues;
  }
};

File: problems/leetcode/0000/54/54.py
# 54. Spiral Matrix
# https://leetcode.com/problems/spiral-matrix/description/

from typing import *

class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        m, n = len(matrix) - 1, len(matrix[0]) - 1
        x, y = 0, 0

        elements = []
        elements.append(matrix[x][y])

        while m > 0 or n > 0:
            for _ in range(n):
                y += 1
                elements.append(matrix[x][y])

            if m == 0:
                break
            for _ in range(m):
                x += 1
                elements.append(matrix[x][y])
            m -= 1

            if n == 0:
                break
            for _ in range(n):
                y -= 1
                elements.append(matrix[x][y])
            n -= 1

            if m == 0:
                break
            for _ in range(m):
                x -= 1
                elements.append(matrix[x][y])
            m -= 1

            if n == 0:
                break
            n -= 1
            y += 1
            elements.append(matrix[x][y])

        return elements
